### PR TITLE
Feature: Device firmware update

### PR DESCRIPTION
**Pull request by svranesevic**
Gradle:
Regenerated gradle wrapper
Incremented version minor

Readme:
Updated version of WolkConnect-Java

Reading:
Renamed to SensorReading

InboundMessageDeserializer:
Handles InboundMessage deserialization

JsonSingleInboundMessageDeserializer:
InboundMessageDeserializer implementation for Json single protocol

Wolk:
Future client extracted to, and replaced by, ConnectivityService/MqttConnectivityService

OutboundMessageFactory:
Extended with OutgoingMessage-s introduced with DFU and file transfer

Firmware update:
Added file transfer via MQTT (Direct file transfer)
Added file transfer from URL, by custom implementation (Url file transfer)
Added firmware update

WolkBuilder:
Added API to allow for configuration of DFU and file transfer used

### CODE REVIEW

**Comment by svranesevic** [comment]
**- Do NOT merge -**
---
**Perform code review only**

File: src/main/java/com/wolkabout/wolk/JsonSingleOutboundMessageFactory.java [hunk 1]
```diff
@@ -43,21 +47,49 @@ public OutboundMessage makeFromReadings(List<Reading> readings) throws IllegalAr
...
+            return new OutboundMessage(payload, channel, alarms.size());
+        } catch (JsonProcessingException e) {
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    @Override
+    public OutboundMessage makeFromFirmwareUpdateStatus(FirmwareUpdateStatus status) throws IllegalArgumentException {
+        try {
+        final String payload = new ObjectMapper().writeValueAsString(status);
```
Comment (nantic) [inline src/main/java/com/wolkabout/wolk/JsonSingleOutboundMessageFactory.java]: Formatting

File: src/main/java/com/wolkabout/wolk/JsonSingleOutboundMessageFactory.java [hunk 2]
```diff
@@ -43,21 +47,49 @@ public OutboundMessage makeFromReadings(List<Reading> readings) throws IllegalAr
...
+        } catch (JsonProcessingException e) {
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    @Override
+    public OutboundMessage makeFromFileTransferPacketRequest(FileTransferPacketRequest request) throws IllegalArgumentException {
+        try {
+            final String payload = new ObjectMapper().writeValueAsString(request);
+            final String channel = "service/status/file/" + deviceKey;
```
Comment (nantic) [inline src/main/java/com/wolkabout/wolk/JsonSingleOutboundMessageFactory.java]: "service/status/file/" should be extracted to a constant.

Comment (nantic) [reply src/main/java/com/wolkabout/wolk/JsonSingleOutboundMessageFactory.java]: Along with all other topics.

File: src/main/java/com/wolkabout/wolk/filetransfer/FileAssembler.java [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+    public enum PacketProcessingError {
+        NONE,
+        OUT_OF_ORDER,
+        INVALID_CHECKSUM,
+        RETRY_COUNT_EXCEEDED,
+        UNABLE_TO_FINALIZE_FILE_CREATION
+    }
+
+    private final Path downloadDirectory;
```
Comment (nantic) [inline src/main/java/com/wolkabout/wolk/filetransfer/FileAssembler.java]: Some of these should be static.

File: src/main/java/com/wolkabout/wolk/filetransfer/FileAssembler.java [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+        } catch (IOException e) {
+            LOG.error("Unable to verify file SHA-256", e);
+            return false;
+        }
+    }
+
+    public void setListener(FileReceiver listener) {
+        this.listener = listener;
+    }
+
```
Comment (nantic) [inline src/main/java/com/wolkabout/wolk/filetransfer/FileAssembler.java]: Empty lines

File: src/main/java/com/wolkabout/wolk/filetransfer/FileAssembler.java [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+            tmpFileOutputStream.write(packet.getData());
+
+            currentPacketId += 1;
+            previousPacketHash = packet.getHash();
+        } catch (IOException e) {
+            LOG.error("Unable to append file chunk to temporary file", e);
+        }
+    }
+
+    private boolean createAndValidateFile() {
```
Comment (nantic) [inline src/main/java/com/wolkabout/wolk/filetransfer/FileAssembler.java]: The method returns boolean but does not follow the naming convention.
Either return void and throw exceptions or rename method.

File: src/main/java/com/wolkabout/wolk/filetransfer/FileTransferPacket.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+    public FileTransferPacket(final byte[] bytes) throws IllegalArgumentException {
+        /*
+         Packet structure:
+         Part 1:   32 bytes   - SHA-256 Hash of previous packet
+         Part 2: >= 1 byte(s) - File chunk
+         Part 3:   32 bytes   - SHA-256 hash of concatenated Part 1 and Part 2
+         */
+
+        if (bytes.length <= 64) {
```
Comment (nantic) [inline src/main/java/com/wolkabout/wolk/filetransfer/FileTransferPacket.java]: What is this 64?

File: src/main/java/com/wolkabout/wolk/firmwareupdate/FirmwareUpdate.java [hunk 7]
```diff
@@ -0,0 +1,443 @@
...
+
+            case INSTALL:
+                handleInstall();
+                break;
+
+            case ABORT:
+                handleAbort();
+                break;
+
+            case UNKNOWN:
```
Comment (nantic) [inline src/main/java/com/wolkabout/wolk/firmwareupdate/FirmwareUpdate.java]: default...

File: src/main/java/com/wolkabout/wolk/firmwareupdate/FirmwareUpdate.java [hunk 8]
```diff
@@ -0,0 +1,443 @@
...
+    }
+
+    public void handlePacket(FileTransferPacket packet) {
+        switch (state) {
+            case PACKET_FILE_TRANSFER:
+                LOG.debug("Processing file transfer packet");
+                onPacketTimeout.cancel(false);
+
+                final FileAssembler.PacketProcessingError error = fileAssembler.processPacket(packet);
+                switch (error) {
```
Comment (nantic) [inline src/main/java/com/wolkabout/wolk/firmwareupdate/FirmwareUpdate.java]: Please refactor so there is no switch in switch.

File: src/main/java/com/wolkabout/wolk/firmwareupdate/FirmwareUpdate.java [hunk 9]
```diff
@@ -0,0 +1,443 @@
...
+                    firmwareFile.toFile().delete();
+                }
+
+                state = State.IDLE;
+                listenerOnStatus(FirmwareUpdateStatus.ok(FirmwareUpdateStatus.StatusCode.ABORTED));
+                break;
+
+            case IDLE:
+                LOG.warn("Ignoring abort command. Reason: Firmware update not initialized");
+                break;
```
Comment (nantic) [inline src/main/java/com/wolkabout/wolk/firmwareupdate/FirmwareUpdate.java]: Missing default.

Comment (nantic) [reply src/main/java/com/wolkabout/wolk/firmwareupdate/FirmwareUpdate.java]: Missing default for all cases.

File: src/main/java/com/wolkabout/wolk/firmwareupdate/FirmwareUpdateStatus.java [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.util.Objects;
+
+public class FirmwareUpdateStatus {
+    public enum StatusCode {
+
+        @JsonProperty("FILE_TRANSFER")
```
Comment (nantic) [inline src/main/java/com/wolkabout/wolk/firmwareupdate/FirmwareUpdateStatus.java]: WTF  is this supposed to do?

Comment (svranesevic) [reply src/main/java/com/wolkabout/wolk/firmwareupdate/FirmwareUpdateStatus.java]: Explicit definition how Jackson should serialize enumeration

File: src/main/java/com/wolkabout/wolk/filetransfer/FileAssembler.java (hunk 3 above)
Comment (svranesevic) [reply src/main/java/com/wolkabout/wolk/filetransfer/FileAssembler.java]: 'LOG' and 'PACKET_SIZE' can be, and are static, rest can not be static

**Pull request merged**